// @workflow/world/dist/queue.js
import { z } from 'zod/v4';
export const QueuePrefix = z.union([
    z.literal('__wkf_step_'),
    z.literal('__wkf_workflow_'),
]);
export const ValidQueueName = z.templateLiteral([QueuePrefix, z.string()]);
export const MessageId = z
    .string()
    .brand()
    .describe('A stored queue message ID');
/**
 * OpenTelemetry trace context for distributed tracing
 */
export const TraceCarrierSchema = z.record(z.string(), z.string());
export const WorkflowInvokePayloadSchema = z.object({
    runId: z.string(),
    traceCarrier: TraceCarrierSchema.optional(),
    requestedAt: z.coerce.date().optional(),
});
export const StepInvokePayloadSchema = z.object({
    workflowName: z.string(),
    workflowRunId: z.string(),
    workflowStartedAt: z.number(),
    stepId: z.string(),
    traceCarrier: TraceCarrierSchema.optional(),
    requestedAt: z.coerce.date().optional(),
});
/**
 * Health check payload - used to verify that the queue pipeline
 * can deliver messages to workflow/step endpoints.
 */
export const HealthCheckPayloadSchema = z.object({
    __healthCheck: z.literal(true),
    correlationId: z.string(),
});
export const QueuePayloadSchema = z.union([
    WorkflowInvokePayloadSchema,
    StepInvokePayloadSchema,
    HealthCheckPayloadSchema,
]);
