// @workflow/core/dist/telemetry/semantic-conventions.js
/**
 * OpenTelemetry semantic conventions for Vercel Workflow telemetry.
 *
 * This module provides standardized telemetry attributes following OpenTelemetry semantic conventions
 * for instrumenting workflow execution, step processing, and related operations. Each exported function
 * creates a properly formatted attribute object that can be used with OpenTelemetry spans.
 *
 * The semantic conventions are organized into several categories:
 * - **Workflow attributes**: Track workflow lifecycle, status, and metadata
 * - **Step attributes**: Monitor individual step execution, retries, and results
 * - **Queue attributes**: Instrument message queue operations
 * - **Deployment attributes**: Capture deployment environment information
 *
 * All attribute functions are type-safe and leverage existing backend types to ensure
 * consistency between telemetry data and actual system state.
 *
 * @example
 * ```typescript
 * import * as Attribute from './telemetry/semantic-conventions.js';
 *
 * // Set workflow attributes on a span
 * span.setAttributes({
 *   ...Attribute.WorkflowName('my-workflow'),
 *   ...Attribute.WorkflowOperation('start'),
 *   ...Attribute.WorkflowRunStatus('running'),
 * });
 *
 * // Set step attributes
 * span.setAttributes({
 *   ...Attribute.StepName('process-data'),
 *   ...Attribute.StepStatus('completed'),
 *   ...Attribute.StepAttempt(1),
 * });
 * ```
 *
 * @see {@link https://opentelemetry.io/docs/specs/semconv/} OpenTelemetry Semantic Conventions
 * @packageDocumentation
 */
/**
 * Creates a semantic convention function that returns an attribute object.
 * @param name - The attribute name following OpenTelemetry semantic conventions
 * @returns A function that takes a value and returns an attribute object
 */
function SemanticConvention(...names) {
    return (value) => Object.fromEntries(names.map((name) => [name, value]));
}
// Workflow attributes
/** The name of the workflow being executed */
export const WorkflowName = SemanticConvention('workflow.name');
/** The operation being performed on the workflow */
export const WorkflowOperation = SemanticConvention('workflow.operation');
/** Unique identifier for a specific workflow run instance */
export const WorkflowRunId = SemanticConvention('workflow.run.id');
/** Current status of the workflow run */
export const WorkflowRunStatus = SemanticConvention('workflow.run.status');
/** Timestamp when the workflow execution started (Unix timestamp) */
export const WorkflowStartedAt = SemanticConvention('workflow.started_at');
/** Number of events processed during workflow execution */
export const WorkflowEventsCount = SemanticConvention('workflow.events.count');
/** Number of arguments passed to the workflow */
export const WorkflowArgumentsCount = SemanticConvention('workflow.arguments.count');
/** Type of the workflow result */
export const WorkflowResultType = SemanticConvention('workflow.result.type');
/** Whether trace context was propagated to this workflow execution */
export const WorkflowTracePropagated = SemanticConvention('workflow.trace.propagated');
/** Name of the error that caused workflow failure */
export const WorkflowErrorName = SemanticConvention('workflow.error.name');
/** Error message when workflow fails */
export const WorkflowErrorMessage = SemanticConvention('workflow.error.message');
/** Number of steps created during workflow execution */
export const WorkflowStepsCreated = SemanticConvention('workflow.steps.created');
/** Number of hooks created during workflow execution */
export const WorkflowHooksCreated = SemanticConvention('workflow.hooks.created');
/** Number of waits created during workflow execution */
export const WorkflowWaitsCreated = SemanticConvention('workflow.waits.created');
// Step attributes
/** Name of the step function being executed */
export const StepName = SemanticConvention('step.name');
/** Unique identifier for the step instance */
export const StepId = SemanticConvention('step.id');
/** Current attempt number for step execution (starts at 1) */
export const StepAttempt = SemanticConvention('step.attempt');
/** Current status of the step */
export const StepStatus = SemanticConvention('step.status');
/** Maximum number of retries allowed for this step */
export const StepMaxRetries = SemanticConvention('step.max_retries');
/** Whether trace context was propagated to this step execution */
export const StepTracePropagated = SemanticConvention('step.trace.propagated');
/** Whether the step was skipped during execution */
export const StepSkipped = SemanticConvention('step.skipped');
/** Reason why the step was skipped */
export const StepSkipReason = SemanticConvention('step.skip_reason');
/** Number of arguments passed to the step function */
export const StepArgumentsCount = SemanticConvention('step.arguments.count');
/** Type of the step result */
export const StepResultType = SemanticConvention('step.result.type');
/** Name of the error that caused step failure */
export const StepErrorName = SemanticConvention('step.error.name');
/** Error message when step fails */
export const StepErrorMessage = SemanticConvention('step.error.message');
/** Whether the step failed with a fatal error (no retries) */
export const StepFatalError = SemanticConvention('step.fatal_error');
/** Whether all retry attempts have been exhausted */
export const StepRetryExhausted = SemanticConvention('step.retry.exhausted');
/** Number of seconds to wait before next retry attempt */
export const StepRetryTimeoutSeconds = SemanticConvention('step.retry.timeout_seconds');
/** Whether the step will be retried after this failure */
export const StepRetryWillRetry = SemanticConvention('step.retry.will_retry');
// Queue attributes
/** Name of the queue being used for message processing */
export const QueueName = SemanticConvention('queue.name');
/** The message id being handled */
export const QueueMessageId = SemanticConvention('messaging.message.id', 'queue.message.id');
/** Time taken to enqueue the message in milliseconds */
export const QueueOverheadMs = SemanticConvention('queue.overhead_ms');
// Deployment attributes
/** Unique identifier for the deployment environment */
export const DeploymentId = SemanticConvention('deployment.id');
// Hook attributes
/** Token identifying a specific hook */
export const HookToken = SemanticConvention('workflow.hook.token');
/** Unique identifier for a hook instance */
export const HookId = SemanticConvention('workflow.hook.id');
/** Whether a hook was found by its token */
export const HookFound = SemanticConvention('workflow.hook.found');
// Webhook attributes
/** Number of webhook handlers triggered */
export const WebhookHandlersTriggered = SemanticConvention('webhook.handlers.triggered');
// Suspension attributes
export const WorkflowSuspensionState = SemanticConvention('workflow.suspension.state');
export const WorkflowSuspensionHookCount = SemanticConvention('workflow.suspension.hook_count');
export const WorkflowSuspensionStepCount = SemanticConvention('workflow.suspension.step_count');
export const WorkflowSuspensionWaitCount = SemanticConvention('workflow.suspension.wait_count');
