// @workflow/builders/dist/standalone.js
import { BaseBuilder } from './base-builder.js';
export class StandaloneBuilder extends BaseBuilder {
    constructor(config) {
        super({
            ...config,
            dirs: ['.'],
        });
    }
    async build() {
        const inputFiles = await this.getInputFiles();
        const tsconfigPath = await this.findTsConfigPath();
        const options = {
            inputFiles,
            tsconfigPath,
        };
        const manifest = await this.buildStepsBundle(options);
        await this.buildWorkflowsBundle(options);
        await this.buildWebhookFunction();
        // Build unified manifest from workflow bundle
        const workflowBundlePath = this.resolvePath(this.config.workflowsBundlePath);
        const manifestDir = this.resolvePath('.well-known/workflow/v1');
        await this.createManifest({
            workflowBundlePath,
            manifestDir,
            manifest,
        });
        await this.createClientLibrary();
    }
    async buildStepsBundle({ inputFiles, tsconfigPath, }) {
        console.log('Creating steps bundle at', this.config.stepsBundlePath);
        const stepsBundlePath = this.resolvePath(this.config.stepsBundlePath);
        await this.ensureDirectory(stepsBundlePath);
        const { manifest } = await this.createStepsBundle({
            outfile: stepsBundlePath,
            inputFiles,
            tsconfigPath,
        });
        return manifest;
    }
    async buildWorkflowsBundle({ inputFiles, tsconfigPath, }) {
        console.log('Creating workflows bundle at', this.config.workflowsBundlePath);
        const workflowBundlePath = this.resolvePath(this.config.workflowsBundlePath);
        await this.ensureDirectory(workflowBundlePath);
        await this.createWorkflowsBundle({
            outfile: workflowBundlePath,
            inputFiles,
            tsconfigPath,
        });
    }
    async buildWebhookFunction() {
        console.log('Creating webhook bundle at', this.config.webhookBundlePath);
        const webhookBundlePath = this.resolvePath(this.config.webhookBundlePath);
        await this.ensureDirectory(webhookBundlePath);
        await this.createWebhookBundle({
            outfile: webhookBundlePath,
        });
    }
}
