// @workflow/next/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withWorkflow = withWorkflow;
const semver_1 = __importDefault(require("semver"));
const builder_js_1 = require("./builder.js");
function withWorkflow(nextConfigOrFn, { workflows, } = {}) {
    if (!process.env.VERCEL_DEPLOYMENT_ID) {
        if (!process.env.WORKFLOW_TARGET_WORLD) {
            process.env.WORKFLOW_TARGET_WORLD = 'local';
            process.env.WORKFLOW_LOCAL_DATA_DIR = '.next/workflow-data';
        }
        const maybePort = workflows?.local?.port;
        if (maybePort) {
            process.env.PORT = maybePort.toString();
        }
    }
    else {
        if (!process.env.WORKFLOW_TARGET_WORLD) {
            process.env.WORKFLOW_TARGET_WORLD = 'vercel';
        }
    }
    return async function buildConfig(phase, ctx) {
        const loaderPath = require.resolve('./loader');
        let nextConfig;
        if (typeof nextConfigOrFn === 'function') {
            nextConfig = await nextConfigOrFn(phase, ctx);
        }
        else {
            nextConfig = nextConfigOrFn;
        }
        // shallow clone to avoid read-only on top-level
        nextConfig = Object.assign({}, nextConfig);
        // configure the loader if turbopack is being used
        if (!nextConfig.turbopack) {
            nextConfig.turbopack = {};
        }
        if (!nextConfig.turbopack.rules) {
            nextConfig.turbopack.rules = {};
        }
        const existingRules = nextConfig.turbopack.rules;
        const nextVersion = require('next/package.json').version;
        const supportsTurboCondition = semver_1.default.gte(nextVersion, 'v16.0.0');
        for (const key of [
            '*.tsx',
            '*.ts',
            '*.jsx',
            '*.js',
            '*.mjs',
            '*.mts',
            '*.cjs',
            '*.cts',
        ]) {
            nextConfig.turbopack.rules[key] = {
                ...(supportsTurboCondition
                    ? {
                        condition: {
                            ...existingRules[key]?.condition,
                            any: [
                                ...(existingRules[key]?.condition.any || []),
                                {
                                    content: /(use workflow|use step)/,
                                },
                            ],
                        },
                    }
                    : {}),
                loaders: [...(existingRules[key]?.loaders || []), loaderPath],
            };
        }
        // configure the loader for webpack
        const existingWebpackModify = nextConfig.webpack;
        nextConfig.webpack = (...args) => {
            const [webpackConfig] = args;
            if (!webpackConfig.module) {
                webpackConfig.module = {};
            }
            if (!webpackConfig.module.rules) {
                webpackConfig.module.rules = [];
            }
            // loaders in webpack apply bottom->up so ensure
            // ours comes before the default swc transform
            webpackConfig.module.rules.push({
                test: /.*\.(mjs|cjs|cts|ts|tsx|js|jsx)$/,
                loader: loaderPath,
            });
            return existingWebpackModify
                ? existingWebpackModify(...args)
                : webpackConfig;
        };
        // only run this in the main process so it only runs once
        // as Next.js uses child processes for different builds
        if (!process.env.WORKFLOW_NEXT_PRIVATE_BUILT &&
            phase !== 'phase-production-server') {
            const shouldWatch = process.env.NODE_ENV === 'development';
            const NextBuilder = await (0, builder_js_1.getNextBuilder)();
            const workflowBuilder = new NextBuilder({
                watch: shouldWatch,
                // discover workflows from pages/app entries
                dirs: ['pages', 'app', 'src/pages', 'src/app'],
                workingDir: process.cwd(),
                buildTarget: 'next',
                workflowsBundlePath: '', // not used in base
                stepsBundlePath: '', // not used in base
                webhookBundlePath: '', // node used in base
                externalPackages: [
                    // server-only and client-only are pseudo-packages handled by Next.js
                    // during its build process. We mark them as external to prevent esbuild
                    // from failing when bundling code that imports them.
                    // See: https://nextjs.org/docs/app/getting-started/server-and-client-components
                    'server-only',
                    'client-only',
                    ...(nextConfig.serverExternalPackages || []),
                ],
            });
            await workflowBuilder.build();
            process.env.WORKFLOW_NEXT_PRIVATE_BUILT = '1';
        }
        return nextConfig;
    };
}
