// @workflow/core/dist/telemetry.js
import { once } from '@workflow/utils';
import { WorkflowSuspension } from './global.js';
import * as Attr from './telemetry/semantic-conventions.js';
// ============================================================
// Trace Context Propagation Utilities
// ============================================================
/**
 * Serializes the current trace context into a format that can be passed through queues
 * @returns A record of strings representing the trace context
 */
export async function serializeTraceCarrier() {
    const otel = await OtelApi.value;
    if (!otel)
        return {};
    const carrier = {};
    // Inject the current context into the carrier
    otel.propagation.inject(otel.context.active(), carrier);
    return carrier;
}
/**
 * Deserializes trace context and returns a context that can be used to continue the trace
 * @param traceCarrier The serialized trace context
 * @returns OpenTelemetry context with the restored trace
 */
export async function deserializeTraceCarrier(traceCarrier) {
    const otel = await OtelApi.value;
    if (!otel)
        return;
    // Extract the context from the carrier
    return otel.propagation.extract(otel.context.active(), traceCarrier);
}
/**
 * Runs a function within the context of a deserialized trace
 * @param traceCarrier The serialized trace carrier (optional)
 * @param fn The function to run within the trace context
 * @returns The result of the function
 */
export async function withTraceContext(traceCarrier, fn) {
    if (!traceCarrier) {
        return fn();
    }
    const otel = await OtelApi.value;
    if (!otel)
        return fn();
    const extractedContext = await deserializeTraceCarrier(traceCarrier);
    if (!extractedContext) {
        return fn();
    }
    return otel.context.with(extractedContext, async () => await fn());
}
const OtelApi = once(async () => {
    try {
        return await import('@opentelemetry/api');
    }
    catch {
        console.warn('OpenTelemetry not available, tracing will be disabled');
        return null;
    }
});
const Tracer = once(async () => {
    const api = await OtelApi.value;
    if (!api)
        return null;
    return api.trace.getTracer('workflow');
});
export async function trace(spanName, ...args) {
    const [tracer, otel] = await Promise.all([Tracer.value, OtelApi.value]);
    const { fn, opts } = typeof args[0] === 'function'
        ? { fn: args[0], opts: {} }
        : { fn: args[1], opts: args[0] };
    if (!fn)
        throw new Error('Function to trace must be provided');
    if (!tracer || !otel) {
        return await fn();
    }
    return tracer.startActiveSpan(spanName, opts, async (span) => {
        try {
            const result = await fn(span);
            span.setStatus({ code: otel.SpanStatusCode.OK });
            return result;
        }
        catch (e) {
            span.setStatus({
                code: otel.SpanStatusCode.ERROR,
                message: e.message,
            });
            applyWorkflowSuspensionToSpan(e, otel, span);
            throw e;
        }
        finally {
            span.end();
        }
    });
}
/**
 * Applies workflow suspension attributes to the given span if the error is a WorkflowSuspension
 * which is technically not an error, but an algebraic effect indicating suspension.
 */
function applyWorkflowSuspensionToSpan(error, otel, span) {
    if (!error || !WorkflowSuspension.is(error)) {
        return;
    }
    span.setStatus({ code: otel.SpanStatusCode.OK });
    span.setAttributes({
        ...Attr.WorkflowSuspensionState('suspended'),
        ...Attr.WorkflowSuspensionStepCount(error.stepCount),
        ...Attr.WorkflowSuspensionHookCount(error.hookCount),
        ...Attr.WorkflowSuspensionWaitCount(error.waitCount),
    });
}
export async function getSpanContextForTraceCarrier(carrier) {
    const [deserialized, otel] = await Promise.all([
        deserializeTraceCarrier(carrier),
        OtelApi.value,
    ]);
    if (!deserialized || !otel)
        return;
    return otel.trace.getSpanContext(deserialized);
}
export async function getActiveSpan() {
    return await withOtel((otel) => otel.trace.getActiveSpan());
}
export function instrumentObject(prefix, o) {
    const handlers = {};
    for (const key of Object.keys(o)) {
        if (typeof o[key] !== 'function') {
            handlers[key] = o[key];
        }
        else {
            const f = o[key];
            // @ts-expect-error
            handlers[key] = async (...args) => trace(`${prefix}.${String(key)}`, {}, () => f(...args));
        }
    }
    return handlers;
}
export async function getSpanKind(field) {
    return withOtel((x) => x.SpanKind[field]);
}
export async function withOtel(fn) {
    const otel = await OtelApi.value;
    if (!otel)
        return undefined;
    return await fn(otel);
}
export function linkToCurrentContext() {
    return withOtel((otel) => {
        const context = otel.trace.getActiveSpan()?.spanContext();
        if (!context)
            return;
        return [{ context }];
    });
}
