// @workflow/world-local/dist/index.js
import { config } from './config.js';
import { initDataDir } from './init.js';
import { createQueue } from './queue.js';
import { createStorage } from './storage.js';
import { createStreamer } from './streamer.js';
// Re-export init types and utilities for consumers
export { DataDirAccessError, DataDirVersionError, ensureDataDir, initDataDir, parseVersion, } from './init.js';
/**
 * Creates a local world instance that combines queue, storage, and streamer functionalities.
 *
 * @param args - Optional configuration object
 * @param args.dataDir - Directory for storing workflow data (default: `.workflow-data/`)
 * @param args.port - Port override for queue transport (default: auto-detected)
 * @param args.baseUrl - Full base URL override for queue transport (default: `http://localhost:{port}`)
 * @throws {DataDirAccessError} If the data directory cannot be created or accessed
 * @throws {DataDirVersionError} If the data directory version is incompatible
 */
export function createLocalWorld(args) {
    const definedArgs = args
        ? Object.fromEntries(Object.entries(args).filter(([, value]) => value !== undefined))
        : {};
    const mergedConfig = { ...config.value, ...definedArgs };
    return {
        ...createQueue(mergedConfig),
        ...createStorage(mergedConfig.dataDir),
        ...createStreamer(mergedConfig.dataDir),
        async start() {
            await initDataDir(mergedConfig.dataDir);
        },
    };
}
