// @workflow/core/dist/runtime/start.js
import { waitUntil } from '@vercel/functions';
import { WorkflowRuntimeError } from '@workflow/errors';
import { withResolvers } from '@workflow/utils';
import { Run } from '../runtime.js';
import { dehydrateWorkflowArguments } from '../serialization.js';
import * as Attribute from '../telemetry/semantic-conventions.js';
import { serializeTraceCarrier, trace } from '../telemetry.js';
import { waitedUntil } from '../util.js';
import { getWorld } from './world.js';
export async function start(workflow, argsOrOptions, options) {
    return await waitedUntil(() => {
        // @ts-expect-error this field is added by our client transform
        const workflowName = workflow?.workflowId;
        if (!workflowName) {
            throw new WorkflowRuntimeError(`'start' received an invalid workflow function. Ensure the Workflow Development Kit is configured correctly and the function includes a 'use workflow' directive.`, { slug: 'start-invalid-workflow-function' });
        }
        return trace(`WORKFLOW.start ${workflowName}`, async (span) => {
            span?.setAttributes({
                ...Attribute.WorkflowName(workflowName),
                ...Attribute.WorkflowOperation('start'),
            });
            let args = [];
            let opts = options ?? {};
            if (Array.isArray(argsOrOptions)) {
                args = argsOrOptions;
            }
            else if (typeof argsOrOptions === 'object') {
                opts = argsOrOptions;
            }
            span?.setAttributes({
                ...Attribute.WorkflowArgumentsCount(args.length),
            });
            const world = opts?.world ?? getWorld();
            const deploymentId = opts.deploymentId ?? (await world.getDeploymentId());
            const ops = [];
            const { promise: runIdPromise, resolve: resolveRunId } = withResolvers();
            const workflowArguments = dehydrateWorkflowArguments(args, ops, runIdPromise);
            // Serialize current trace context to propagate across queue boundary
            const traceCarrier = await serializeTraceCarrier();
            const runResponse = await world.runs.create({
                deploymentId: deploymentId,
                workflowName: workflowName,
                input: workflowArguments,
                executionContext: { traceCarrier },
            });
            resolveRunId(runResponse.runId);
            waitUntil(Promise.all(ops).catch((err) => {
                // Ignore expected client disconnect errors (e.g., browser refresh during streaming)
                const isAbortError = err?.name === 'AbortError' || err?.name === 'ResponseAborted';
                if (!isAbortError)
                    throw err;
            }));
            span?.setAttributes({
                ...Attribute.WorkflowRunId(runResponse.runId),
                ...Attribute.WorkflowRunStatus(runResponse.status),
                ...Attribute.DeploymentId(deploymentId),
            });
            await world.queue(`__wkf_workflow_${workflowName}`, {
                runId: runResponse.runId,
                traceCarrier,
            }, {
                deploymentId,
            });
            return new Run(runResponse.runId);
        });
    });
}
